### ISSUE TITLE
no-magic-numbers and JSX

### ISSUE THREAD

**Issue by oliviertassinari** [eslint/eslint#4185]
When running `no-magic-numbers` on this:

```js
React.createClass({
	render() {
		return <textarea rows={4} />;
	}
});
```

The rule throws for `rows={4}`.
I think that JSX properties are explicit, not magic, hence, they should be ignored.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by hzoo** [comment]
I guess so since the name of it is already there - in your case `rows`.

**Comment by oliviertassinari** [comment]
I will try to submit a PR for this one.

**Comment by existentialism** [comment]
@oliviertassinari if you haven't started already, mind if I submit a PR?

**Comment by oliviertassinari** [comment]
@existentialism. Sure, go ahead

**Comment by nzakas** [comment]
This seems like the correct behavior of this rule. If you want to use a literal here, I'd say put it directly into a JSX attribute:

```
React.createClass({
    render() {
        return <textarea rows="4" />;
    }
});
```

Anything inside of `{` and `}` is considered a JavaScript expression, and therefore,  your `4` is a magic number.

**Comment by aruberto** [comment]
Only issue with using rows="4" is when the component declares attribute rows with PropType number:

http://codepen.io/anon/pen/XmVmMX

In the console, you'll see 

```
Warning: Failed propType: Invalid prop `rows` of type `string` supplied to `MyComponent`, expected `number`.
```

**Comment by nzakas** [comment]
That's a React-specific behavior, and we don't build in library-specific concerns as a general rule.

**Comment by hzoo** [comment]
Can add it to https://github.com/yannickcr/eslint-plugin-react

**Comment by jetpacmonkey** [comment]
How is this any different from
```javascript
var config = {
    rows: 4
};
```
?

**Comment by nzakas** [comment]
That would require an entirely new rule that does almost the same thing. That seems like a worse case scenario situation.

**Comment by nzakas** [comment]
For anyone just stopping by, the question at hand is why this pattern should *not* be considered a magic number:

```js
React.createClass({
    render() {
        return <textarea rows={4} />;
    }
});
```

The 4 is currently considered magic. If you have a good argument as to why it's not, please share.

**Comment by hzoo** [comment]
I think if it's just a number it could be ok since the attribute is right there to tell you that the `rows` are 4?

I don't think doing something like `const numRows = 4` and `rows={numRows}` would be that much better

**Comment by charlesbdudley** [comment]
From the docs: 
> 'Magic numbers' are numbers that occur multiple time in code without an explicit meaning. They should preferably be replaced by named constants.
>http://eslint.org/docs/rules/no-magic-numbers.html

If the 4 is being used multiple times in this component then perhaps it should be replaced by a named constant.

**Comment by chilts** [comment]
To me, `<textarea rows={4} />` is a magic number. However, `<textarea rows="4" />` wouldn't be.

As backup, when converted to JS, it would be translated as `React.createElement("textarea", {rows : 4})` so if that was regular code (which it is), then it would be a magic number.

**Comment by aaronjensen** [comment]
@chilts it looks like there's a setting for that, though I don't understand the docs:

> detectObjects
> A boolean to specify if we should detect numbers when setting object properties for example. false by default.

It's false by default, does that mean that `{x: 4}` is an error or not by default? I agree that it should behave however `{x: 4}` behaves.

**Comment by Spy-Seth** [comment]
> To me, `<textarea rows={4} />` is a magic number. However, `<textarea rows="4" />` wouldn't be.

I cannot see why `<textarea rows="4" />` isn't a magic number. Given to the definition, this have no more meaning than the magic version `<textarea rows={4} />`. 

> 'Magic numbers' are numbers that occur multiple time in code without an explicit meaning.

**Comment by nmn** [comment]
as a react/jsx user, I just want to say that `<textara rows="4" />` is a string not a number. So `<textara rows={4} />` is the only option here. That said, I'm not sure if it counts as a magic number.

Here's a question, is this a magic number?
```
someFunction(4)
```

if so, than so should `<textara rows={4} />`.
Jsx should be treated just like function calls.

**Comment by Spy-Seth** [comment]
>  I just want to say that <textara rows="4" /> is a string not a number. So <textara rows={4} /> is the only option here

Good point!

**Comment by petetnt** [comment]
I think the `propTypes`-problem is an React issue (by design?). The spec says that `rows` value must be an `positive integer`. React requires that JSX values should be either an expression or a quoted JSX text. However, both of the following HTML's are considered valid:

``` 
<textarea cols=1></textarea>
<textarea cols="1"></textarea>
```

out of which the first one isn't allowed by React.

But that's an React issue, as far as this issue goes I don't think that there's any case where `{1}` shouldn't be considered a magic number. If you need to use it as an positive integer or if you reuse the value, use an variable

```
let val = 4;
<textarea cols={val}></textarea>
``` 

If you just want 4 cols, use the string version

```
<textarea cols="4"></text></textarea>
```

If you need propType validation in either case, use `React.PropTypes.oneOfType`


```
optionalUnion: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.number
])
```

**Comment by nzakas** [comment]
@nmn yes, that's considered a magic number by this rule. 

@aaronjensen that logic is specific to React's current implementation, not to JSX on its own.

**Comment by aaronjensen** [comment]
Good point. In that case, it's simplest to treat it as a magic number. If eslint-plugin-react wants to override that behavior, that's its prerogative I'd say.

**Comment by existentialism** [comment]
Makes sense, I'll close the PR!

**Comment by nzakas** [comment]
Thanks for the discussion everyone.

**Comment by oliviertassinari** [comment]
I'm almost convinced now. Thanks everybody
